# Day 2 EOD: A Data Biography of Toolik Field Station Weather

Date: 09/01/2026

In [6]:
import pandas as pd

In [4]:
url = 'https://eds-217-essential-python.github.io/data/toolik_weather.csv'

In [7]:
df = pd.read_csv(url)

# Part 1: Vital Statistics

### 1
How many rows and how many columns does the dataset have?

In [14]:
df.shape

(11171, 21)

This data set has 21 columns and 11,171 rows.

### 2
What are the column names? Get them as a Python list.

In [15]:
df.columns

Index(['Year', 'Month', 'Date', 'LTER_Site', 'Station', 'Daily_AirTemp_Mean_C',
       'Flag_Daily_AirTemp_Mean_C', 'Daily_AirTemp_AbsMax_C',
       'Flag_Daily_AirTemp_AbsMax_C', 'Daily_AirTemp_AbsMin_C',
       'Flag_Daily_AirTemp_AbsMin_C', 'Daily_Precip_Total_mm',
       'Flag_Daily_Precip_Total_mm', 'Daily_windsp_mean_msec',
       'FLAG_Daily_windsp_mean_msec', 'Daily_Windspeed_AbsMax_m_s',
       'Daily_globalrad_total_jcm2', 'FLAG_Daily_globalrad_total_mjm2', 'Moss',
       'Soil20cm', 'Comments'],
      dtype='object')

In [16]:
column_names = df.columns.tolist()

In [17]:
column_names

['Year',
 'Month',
 'Date',
 'LTER_Site',
 'Station',
 'Daily_AirTemp_Mean_C',
 'Flag_Daily_AirTemp_Mean_C',
 'Daily_AirTemp_AbsMax_C',
 'Flag_Daily_AirTemp_AbsMax_C',
 'Daily_AirTemp_AbsMin_C',
 'Flag_Daily_AirTemp_AbsMin_C',
 'Daily_Precip_Total_mm',
 'Flag_Daily_Precip_Total_mm',
 'Daily_windsp_mean_msec',
 'FLAG_Daily_windsp_mean_msec',
 'Daily_Windspeed_AbsMax_m_s',
 'Daily_globalrad_total_jcm2',
 'FLAG_Daily_globalrad_total_mjm2',
 'Moss',
 'Soil20cm',
 'Comments']

The column names are:
- Year
- Month
- Date
- LTER_Site
- Station
- Daily_AirTemp_Mean_C
- Flag_Daily_AirTemp_Mean_C
- Daily_AirTemp_AbsMax_C
- Flag_Daily_AirTemp_AbsMax_C
- Daily_AirTemp_AbsMin_C
- Flag_Daily_AirTemp_AbsMin_C
- Daily_Precip_Total_mm
- Flag_Daily_Precip_Total_mm
- Daily_windsp_mean_msec
- FLAG_Daily_windsp_mean_msec
- Daily_Windspeed_AbsMax_m_s
- Daily_globalrad_total_jcm2
- FLAG_Daily_globalrad_total_mjm2
- Moss
- Soil20cm
- Comments

### 3
What is the data type of each column?

In [20]:
df.dtypes

Year                                 int64
Month                                int64
Date                                 int64
LTER_Site                           object
Station                             object
Daily_AirTemp_Mean_C               float64
Flag_Daily_AirTemp_Mean_C           object
Daily_AirTemp_AbsMax_C             float64
Flag_Daily_AirTemp_AbsMax_C         object
Daily_AirTemp_AbsMin_C             float64
Flag_Daily_AirTemp_AbsMin_C         object
Daily_Precip_Total_mm              float64
Flag_Daily_Precip_Total_mm          object
Daily_windsp_mean_msec             float64
FLAG_Daily_windsp_mean_msec         object
Daily_Windspeed_AbsMax_m_s         float64
Daily_globalrad_total_jcm2         float64
FLAG_Daily_globalrad_total_mjm2     object
Moss                               float64
Soil20cm                           float64
Comments                            object
dtype: object

Year, month, and date are int type columns.  Daily_AirTemp_Mean_c, Daily_AirTemp_absMax_c, Daily_AirTemp_absMin_c, Daily_precip_total_mm, daily_windsp_mean_msec, daily_windspeed_absmax, daily_globalrad_total_j, Moss, and Soil20cm are all floats.  Every other column is an object.

### 4
Yesterday’s practice told you this record runs from June 1988 to December 2018. Do not take that on trust. Use .min() and .max() on the Date column to find the exact first and last day and report those in a new cell.

In [21]:
df['Date'].max()

20181231

In [22]:
df['Date'].min()

19880601

The data runs from 06/01/1988 to 12/31/2018

# Part 2: The Health Check

### 1
Run .info(). How many columns have fewer non-null values than there are rows?

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11171 entries, 0 to 11170
Data columns (total 21 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Year                             11171 non-null  int64  
 1   Month                            11171 non-null  int64  
 2   Date                             11171 non-null  int64  
 3   LTER_Site                        11171 non-null  object 
 4   Station                          11171 non-null  object 
 5   Daily_AirTemp_Mean_C             11171 non-null  float64
 6   Flag_Daily_AirTemp_Mean_C        1310 non-null   object 
 7   Daily_AirTemp_AbsMax_C           11001 non-null  float64
 8   Flag_Daily_AirTemp_AbsMax_C      1020 non-null   object 
 9   Daily_AirTemp_AbsMin_C           10967 non-null  float64
 10  Flag_Daily_AirTemp_AbsMin_C      1236 non-null   object 
 11  Daily_Precip_Total_mm            10751 non-null  float64
 12  Flag_Daily_Precip_

15 columns have fewer non-null values than there are rows (11171)

### 2
Use .isnull().sum() to count missing values per column. Which three columns are the emptiest?

In [26]:
df.isnull().sum()

Year                                   0
Month                                  0
Date                                   0
LTER_Site                              0
Station                                0
Daily_AirTemp_Mean_C                   0
Flag_Daily_AirTemp_Mean_C           9861
Daily_AirTemp_AbsMax_C               170
Flag_Daily_AirTemp_AbsMax_C        10151
Daily_AirTemp_AbsMin_C               204
Flag_Daily_AirTemp_AbsMin_C         9935
Daily_Precip_Total_mm                420
Flag_Daily_Precip_Total_mm          8002
Daily_windsp_mean_msec               826
FLAG_Daily_windsp_mean_msec        11170
Daily_Windspeed_AbsMax_m_s           846
Daily_globalrad_total_jcm2          7053
FLAG_Daily_globalrad_total_mjm2    11156
Moss                                 838
Soil20cm                             820
Comments                            2247
dtype: int64

FLAG_daily_windsp_mean is almost completely empty (i.e. has 11170 null values).  The other most empty columns are FLAG_daily_globalrad_total_j (11156 null values) and Flag_daily_airtemp_absMax (10151 null values).

### 3
One of the flag columns is empty in all but a very small number of rows. Which one, and how many values does it actually have?

FLAG_daily_globalrad_total_j is empty in all but 15 rows

# Part 3: What varies, and what doesn’t

### 1
Use .value_counts() on LTER_Site and on Station. What does each column tell you?

In [27]:
df['LTER_Site'].value_counts()

LTER_Site
ARC    11171
Name: count, dtype: int64

In [28]:
df['Station'].value_counts()

Station
TLKMAIN    11171
Name: count, dtype: int64

.value_counts() tells us that each of these columns is full (contains 11171 values)

### 2
Use .value_counts() on Flag_Daily_AirTemp_Mean_C. What values appear, and what do you think they mean?

In [29]:
df['Flag_Daily_AirTemp_Mean_C'].value_counts()

Flag_Daily_AirTemp_Mean_C
E    1310
Name: count, dtype: int64

This column only has 1310 values

### 3
Use .value_counts() on the Year column and look at both ends of the result, with .head() and then with .tail(). Do all years have the same number of observations? What would explain a year with fewer?

In [32]:
df['Year'].value_counts().head()

Year
1992    366
2000    366
1996    366
2012    366
2016    366
Name: count, dtype: int64

In [33]:
df['Year'].value_counts().tail()

Year
2011    365
2017    365
2015    365
2018    365
1988    214
Name: count, dtype: int64

Some years have 365 results and some have 366 (leap years), since data is collected daily.  One year (1988) has only 214 values, since that is the year that data collection began (on June 1).

# Part 4: The numbers

### 1 
What is the mean daily air temperature across the whole record? The minimum? The maximum?

In [38]:
print(f"Mean: {df['Daily_AirTemp_Mean_C'].mean()}")
print(f"Max: {df['Daily_AirTemp_Mean_C'].max()}")
print(f"Min: {df['Daily_AirTemp_Mean_C'].min()}")


Mean: -8.19643720347328
Max: 20.7
Min: -55.6


The mean daily air temperature across the whole record is -8.2 degrees celsius.  The maximum was 20.7 degrees celsius, while the minimum was -55.6 degrees celsius.

### 2
Are the mean, the minimum and the maximum plausible for a site above the Arctic Circle? Say why or why not.

These are plausible.  The maximum is close to 70 degrees F, which seems a little unlikely but not impossible.  The minimum and average are quite plausible for the arctic.

### 3
Pick one other numeric column and report its range. Does anything about it look wrong?

In [39]:
print(f"Mean: {df['Daily_Precip_Total_mm'].mean()}")
print(f"Max: {df['Daily_Precip_Total_mm'].max()}")
print(f"Min: {df['Daily_Precip_Total_mm'].min()}")

Mean: 0.9509533996837503
Max: 64.5
Min: 0.0


The maximum is 64.5mm, the minimum is 0mm, and the mean is 0.95mm.  This seems plausible, since the max could be a severe snowstorm and the minimum could be a day with no precipitation.

# Part 5: A tidier table

### 1
Build a list of the five columns you’d keep if you had to hand this dataset to a colleague who only cared about temperature. Use it to make a smaller DataFrame.

In [41]:
temp_columns = [
    'Date',
    'Daily_AirTemp_Mean_C',
    'Daily_AirTemp_AbsMax_C',
    'Daily_AirTemp_AbsMin_C',
    'Comments'
]
temp_data = df[temp_columns]

I would include the following columns:
    'Date',
    'Daily_AirTemp_Mean_C',
    'Daily_AirTemp_AbsMax_C',
    'Daily_AirTemp_AbsMin_C',
    'Comments'

### 2
Build a dictionary that renames at least two of those columns to something shorter and clearer. Apply it with .rename(columns=...) and confirm the new names.

In [45]:
new_names = {'Date':'date', 'Daily_AirTemp_Mean_C':'mean_temp', 'Daily_AirTemp_AbsMax_C':'max_temp', 'Daily_AirTemp_AbsMin_C':'min_temp', 'Comments':'comments'}
new_names
temp_data.rename(columns = new_names)

,date,mean_temp,max_temp,min_temp,comments
0,19880601,8.4,NaN,NaN,Air temp 1 & 5 meter estimated from regressio...
1,19880602,6.0,NaN,NaN,Air temp 1 & 5 meter estimated from regressio...
2,19880603,5.8,NaN,NaN,Air temp 1 & 5 meter estimated from regressio...
3,19880604,1.8,NaN,NaN,Air temp 1 & 5 meter estimated from regressio...
4,19880605,6.8,NaN,NaN,Air temp 1 & 5 meter estimated from regressio...
...,...,...,...,...,...
11166,20181227,-21.2,-7.7,-26.7,NaN
11167,20181228,-24.2,-21.4,-26.3,NaN
11168,20181229,-21.1,-16.3,-25.8,NaN
11169,20181230,-17.3,-7.9,-28.0,NaN


# Part 6: Write the biography

This dataset includes weather data measured at the Toolik LTER site in the Arctic circle between 1988 and 2018.  The data begins at 06/01/1988 and ends at 12/31/2018.  The dataset include temperature mean/min/max data, windspeed data, precipitation measurements, and radiation measurements as well as soil and moss data.